### PR TITLE
Limit reviews upto date specified.

### PR DESCRIPTION
**Pull request by scorpion35**
Issue: https://github.com/fastlane/fastlane/issues/13030

Solution: Added an optional `upto_date` parameter to `get_review` methods. By limiting `get_reviews` upto a certain date, the time to return from the method is reduced from hours to seconds!

Examples - 
**Before:** Time taken to get **all** reviews: ~ upto 2 hours (best case). Up to 5 hours (worst case), and the script crashes in the end. 
```
2018-08-03 11:14:33 PM -04 EDT - INFO - [apple_reviews] | Finding all available reviews for "bundleid1" ...
2018-08-04 02:18:39 AM -04 EDT - INFO - [apple_reviews] | Found 90,000 reviews.
...
2018-08-03 11:14:33 PM -04 EDT - INFO - [apple_reviews] | Finding all available reviews for "bundleid2" ...
2018-08-04 02:18:39 AM -04 EDT - ERROR - [apple_reviews] | It seems the script crashed in iteration 1 while getting reviews for App: "bundleid2".
```

**After:** Time taken to get reviews up to a specified date: ~ 2 secs (best case) to ~ 30 secs. 
```
2018-08-05 12:41:00 PM -04 EDT - INFO - [apple_reviews] | Finding all available reviews ...
2018-08-05 12:41:02 PM -04 EDT - INFO - [apple_reviews] | 	Found 100 reviews upto date "2018-07-01"
...
2018-08-05 12:40:12 PM -04 EDT - INFO - [apple_reviews] | Finding all available reviews ...
2018-08-05 12:40:38 PM -04 EDT - INFO - [apple_reviews] | 	Found 1700 reviews upto date "2018-07-01"
```

### CODE REVIEW

**Review by joshdholtz — CHANGES_REQUESTED** [review]
Thanks for making this PR 😊 I probably left too many comments but date stuff is weird 😊 

BTW, we probably **also** want to filter `all_reviews` with the up to the date specified in the parameter. Right now the date logic is only breaking the loop but not filtering the any that _may_ actually be before the date 🙃

File: spaceship/lib/spaceship/tunes/tunes_client.rb [hunk 1]
```diff
@@ -371,6 +371,12 @@ def get_reviews(app_id, platform, storefront, version_id)
 
         r = request(:get, rating_url)
         all_reviews.concat(parse_response(r, 'data')['reviews'])
+
+        last_review_date = Date.parse(Time.at(all_reviews[-1]['value']['lastModified']/1000).to_s).to_date
+        if last_review_date < DateTime.parse(upto_date).to_date
```
[the remark this answers is not in the document]
Comment (scorpion35) [reply spaceship/lib/spaceship/tunes/tunes_client.rb]: Done! https://github.com/fastlane/fastlane/pull/13040/commits/e846dd1ffc61eef579b017400f7dcfd32c80554d#diff-64843dee7bf6cf0342bd0e57d9b43c2b

File: spaceship/lib/spaceship/tunes/app_ratings.rb [hunk 2]
```diff
@@ -46,8 +46,8 @@ def average_rating
       end
 
       # @return (Array) of Review Objects
-      def reviews(store_front = '', version_id = '')
-        raw_reviews = client.get_reviews(application.apple_id, application.platform, store_front, version_id)
+      def reviews(store_front = '', version_id = '', upto_date = '')
```
[the remark this answers is not in the document]
Comment (scorpion35) [reply spaceship/lib/spaceship/tunes/app_ratings.rb]: Good idea. Done! https://github.com/fastlane/fastlane/pull/13040/commits/e846dd1ffc61eef579b017400f7dcfd32c80554d#diff-ba47141cdb85eda8d52ad6ea553966fb

File: spaceship/lib/spaceship/tunes/tunes_client.rb [hunk 3]
```diff
@@ -358,7 +358,7 @@ def get_ratings(app_id, platform, version_id = '', storefront = '')
       parse_response(r, 'data')
     end
 
-    def get_reviews(app_id, platform, storefront, version_id)
+    def get_reviews(app_id, platform, storefront, version_id, upto_date)
```
[the remark this answers is not in the document]
Comment (scorpion35) [reply spaceship/lib/spaceship/tunes/tunes_client.rb]: Done! https://github.com/fastlane/fastlane/pull/13040/commits/e846dd1ffc61eef579b017400f7dcfd32c80554d#diff-ba47141cdb85eda8d52ad6ea553966fb

File: spaceship/lib/spaceship/tunes/tunes_client.rb [hunk 4]
```diff
@@ -371,6 +371,12 @@ def get_reviews(app_id, platform, storefront, version_id)
 
         r = request(:get, rating_url)
         all_reviews.concat(parse_response(r, 'data')['reviews'])
+
+        last_review_date = Date.parse(Time.at(all_reviews[-1]['value']['lastModified']/1000).to_s).to_date
```
[the remark this answers is not in the document]
Comment (scorpion35) [reply spaceship/lib/spaceship/tunes/tunes_client.rb]: Done! https://github.com/fastlane/fastlane/pull/13040/commits/e846dd1ffc61eef579b017400f7dcfd32c80554d#diff-64843dee7bf6cf0342bd0e57d9b43c2b

Please take a look and let me know if you'd like to change it to something else.

**Comment by scorpion35** [comment]
@joshdholtz Have made changes suggested by you! Kindly take a look and let me know if you need me to do anything else.

File: spaceship/lib/spaceship/tunes/tunes_client.rb [hunk 5]
```diff
@@ -371,13 +374,20 @@ def get_reviews(app_id, platform, storefront, version_id)
 
         r = request(:get, rating_url)
         all_reviews.concat(parse_response(r, 'data')['reviews'])
+        last_review_date = Time.at(all_reviews[-1]['value']['lastModified']/1000)
+        
+        break if last_review_date < upto_date && !upto_date.nil?
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/tunes/tunes_client.rb]: I would probably make this break and filter here something like this...
```rb
if upto_date && last_review_date < upto_date
  all_reviews = all_reviews.select { |review| Time.at(review['value']['lastModified']/1000) > upto_date  }
  break
end
```

And then remove the multiple returns at the bottom

File: spaceship/lib/spaceship/tunes/tunes_client.rb [hunk 6]
```diff
@@ -371,13 +374,20 @@ def get_reviews(app_id, platform, storefront, version_id)
...
           index += per_page
         else
           break
         end
       end
-      all_reviews
+
+      return all_reviews if upto_date.nil?
+
+      return all_reviews.select { |review| Time.at(review['value']['lastModified']/1000) > upto_date  }
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/tunes/tunes_client.rb]: Remove these multiple `return`s and just do... (see above comment)

```rb
return all_reviews
```

**Review by joshdholtz — CHANGES_REQUESTED** [review]
Got a few more small comments! It also looks like the tests are failing and so is the linting. So if you could also ...

1) Fix existing tests
2) Add new tests that use the new parameter
3) Run rubocop (the linter) by running `rubocop -a && rubocop -D`
4) Sign CLA

Looking forward to getting this merged after these steps 😊

File: spaceship/lib/spaceship/tunes/tunes_client.rb (hunk 5 above)
Comment (scorpion35) [reply spaceship/lib/spaceship/tunes/tunes_client.rb]: haha cool. I originally had something similar to this, but then I thought I should keep it a one-liner (looked at a couple of other `break` statements). Changed it to like you suggested!
https://github.com/fastlane/fastlane/pull/13040/commits/7c572e21ba96a57a85b0ff75cc0c8853aaaaa48b

File: spaceship/lib/spaceship/tunes/tunes_client.rb (hunk 6 above)
Comment (scorpion35) [reply spaceship/lib/spaceship/tunes/tunes_client.rb]: Done! :)

**Comment by scorpion35** [comment]
@joshdholtz I will add tests tomorrow and look into fixing the linting as well!

**Comment by scorpion35** [comment]
I signed it!

**Comment by scorpion35** [comment]
I signed it!

**Comment by scorpion35** [comment]
@joshdholtz Fixed the linting test. 

I tried to write new tests, but couldn't seem to figure it out. https://github.com/fastlane/fastlane/pull/13040/commits/c57e713a1e57673bddd60b9f39b0314c79157e97 & https://github.com/fastlane/fastlane/pull/13040/commits/ee69cfec35c78c50aadff67e630fca98938b97d0

Could you give me some guidance on how to make a new test, please? I have tried to place a `binding.pry` in the new test, but the test seems to be stuck forever; not sure what I am doing wrong. Any help is greatly appreciated!

**Comment by joshdholtz** [comment]
@scorpion35 Yup, I can help! I'm not sure if I'll have time today but I will look tomorrow/Thursday. Are you cool if I push commits to this PR with tests?

**Comment by scorpion35** [comment]
@joshdholtz Any update, please? :)

**Comment by scorpion35** [comment]
@joshdholtz Any help, please? :) Thanks in advance!

**Comment by scorpion35** [comment]
@joshdholtz @taquitos @KrauseFx Kindly let me know how to add tests to Fastlane, please. I tried to add them but wasn't able to figure it out. If you can give me some tips on how to add them, I will try again.

I am trying to move my script to a remote machine but would like to avoid configuring the remote machine to use my custom changes, would prefer to have the machine automatically update to latest version of Fastlane available! :)

**Review by taquitos — APPROVED** [review]
Lint it up, then everything looks great!
🎈:goat:

**Review by joshdholtz — APPROVED** [review]

**Comment by joshdholtz** [comment]
@scorpion35 Merged! Thanks for making this changing 😊

**Comment by scorpion35** [comment]
You are the best, thank you very much @joshdholtz  :)

It seems I over thought about these tests, almost had it (https://github.com/fastlane/fastlane/commit/ee69cfec35c78c50aadff67e630fca98938b97d0), but over complicated it with some other changes :D haha. 

Thanks again! 🙇 Do you know when can I expect these changes in official Fastlane gem?

**Pull request merged**